### PR TITLE
New `this_cpu` APIs

### PR DESCRIPTION
**Pull request by junyang-zh**
A part of #1158 , I decide to separate it from that PR and discuss about the API design of `this_cpu` here.

### CODE REVIEW

File: ostd/src/cpu/mod.rs [hunk 1]
```diff
@@ -11,43 +11,106 @@ cfg_if::cfg_if! {
...
+impl Deref for ThisCpu<'_> {
+    type Target = u32;
+
+    fn deref(&self) -> &Self::Target {
+        &self.id
+    }
+}
+
+enum ThisCpuGuard<'a> {
+    Owning(DisablePreemptGuard),
```
Comment (tatetian) [inline ostd/src/cpu/mod.rs:88]: The word `Owned` is better as it is consistent with Rust std's `Cow`.

File: kernel/aster-nix/src/sched/priority_scheduler.rs [hunk 2]
```diff
@@ -71,13 +72,19 @@ impl<T: Sync + Send + PreemptSchedInfo> Scheduler<T> for PreemptScheduler<T> {
     }
 
     fn local_rq_with(&self, f: &mut dyn FnMut(&dyn LocalRunQueue<T>)) {
-        let local_rq: &PreemptRunQueue<T> = &self.rq[this_cpu() as usize].disable_irq().lock();
+        let preempt_guard = disable_preempt();
+        let local_rq: &PreemptRunQueue<T> = &self.rq[*this_cpu_with(&preempt_guard) as usize]
+            .disable_irq()
```
Comment (tatetian) [inline kernel/aster-nix/src/sched/priority_scheduler.rs:77]: The preemption guard is redundant to the subsequent IRQ guard. If `this_cpu_with` can be implemented as a generic function over `G`, which can be either `PreemptGuard` or `IrqDisalbedGuard`, then the user code here avoids taking the preemption guard.

```rust
let irq_disabled_guard = disable_irq();
let this_cpu = this_cpu_with(&irq_disabled_guard);
let local_rq = &self.rq[*this_cpu].lock();
f(local_rq)
```

Similarly, the `CpuLocal::borrow_with` can also benefit from being generic over `G`, which represents the guard type.

All we have to do is to introduce a new marker trait and use it in the two functions.

```rust
// file: cpu/local.rs

/// A marker trait for guard types that can "pin" the current task to the current CPU.
/// Such guard types include `DisabledLocalIrqGuard` and `DisablePreemptGuard`.
/// Only when the current task is being pinned can CPU-local variables be stable.
pub unsafe trait PinCurrentTask {}

unsafe impl PinCurrentTask for DisabledLocalIrqGuard {}
unsafe impl PinCurrentTask for DisablePreemptGuard {}
```

**Comment by junyang-zh** [comment]
@boterinas rerun

**Comment by tatetian** [comment]
@boterinas rerun

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**